File: python/04-prompts/few-shot.py
"""Comparación 0-shot vs 3-shot vs 6-shot en clasificación de sentimiento.

Demuestra:
- Cómo el número de ejemplos afecta accuracy y consistencia del formato
- Majority label bias: con ejemplos desbalanceados, el modelo predice la clase dominante
- Recency bias: el último ejemplo del prompt influye desproporcionadamente
- Los tokens consumidos por cada variante

Cómo ejecutar:
    make py SCRIPT=python/04-prompts/few-shot.py

Qué esperar:
    Comparacion de accuracy con 0, 3 y 6 ejemplos. Demostracion de majority
    label bias y recency bias con ejemplos deliberadamente desbalanceados.

Variables de entorno:
    MODEL — modelo a usar (default: claude-sonnet-4-6)
"""

import os
import json
import re
import anthropic

MODEL = os.environ.get("MODEL", "claude-sonnet-4-6")

# ─── 1. Dataset de prueba ───────────────────────────────────────────────────

REVIEWS = [
    {"text": "El producto llegó perfectamente empaquetado y funciona exactamente como se describe. Muy satisfecho.", "label": "positivo"},
    {"text": "Terrible experiencia. El artículo llegó roto y el servicio de atención al cliente no respondió.", "label": "negativo"},
    {"text": "El producto cumple lo básico. No es el mejor que he usado pero tampoco es malo. Precio razonable.", "label": "neutro"},
    {"text": "¡Increíble calidad! Superó todas mis expectativas. Lo recomendaría sin dudar.", "label": "positivo"},
    {"text": "Llegó tarde y el embalaje estaba dañado. El producto funciona pero la experiencia de compra fue mala.", "label": "negativo"},
]

# ─── 2. Ejemplos few-shot ───────────────────────────────────────────────────

EXAMPLES_3_SHOT = [
    {"text": "Excelente producto, muy buena calidad y entrega rápida.", "label": "positivo"},
    {"text": "No me gustó nada. Tuve que devolverlo al día siguiente.", "label": "negativo"},
    {"text": "Hace lo que promete. Ni más ni menos.", "label": "neutro"},
]

EXAMPLES_6_SHOT = EXAMPLES_3_SHOT + [
    {"text": "Mejor compra del año, totalmente recomendado para todos.", "label": "positivo"},
    {"text": "Producto defectuoso. Una pérdida de dinero total.", "label": "negativo"},
    {"text": "Está bien para lo que cuesta. No hay mucho que decir.", "label": "neutro"},
]


def build_examples_block(examples: list[dict]) -> str:
    """Construye el bloque XML de ejemplos few-shot."""
    lines = ["<examples>"]
    for ex in examples:
        lines.append("  <example>")
        lines.append(f'    <texto>{ex["text"]}</texto>')
        lines.append(f'    <sentimiento>{ex["label"]}</sentimiento>')
        lines.append("  </example>")
    lines.append("</examples>")
    return "\n".join(lines)


def build_system_prompt(examples: list[dict]) -> str:
    """Construye el system prompt con los ejemplos dados."""
    base = (
        "Clasifica el sentimiento de reseñas de producto como: positivo, negativo o neutro.\n"
        "Responde SOLO con una de estas tres palabras: positivo, negativo, neutro.\n"
        "Sin explicaciones adicionales.\n"
    )
    if not examples:
        return base
    return base + "\n" + build_examples_block(examples)


# ─── 3. Clasificación ───────────────────────────────────────────────────────

def classify_reviews(client: anthropic.Anthropic, system_prompt: str, reviews: list[dict]) -> list[dict]:
    """Clasifica una lista de reseñas y devuelve resultados con métricas."""
    results = []
    for review in reviews:
        response = client.messages.create(
            model=MODEL,
            max_tokens=20,
            system=system_prompt,
            messages=[{"role": "user", "content": review["text"]}],
        )
        raw_output = response.content[0].text.strip().lower()
        # Normalizar: extraer la primera palabra que sea una de las tres clases
        match = re.search(r"\b(positivo|negativo|neutro)\b", raw_output)
        predicted = match.group(1) if match else raw_output

        results.append({
            "text": review["text"][:50] + "...",
            "label_real": review["label"],
            "prediccion": predicted,
            "correcto": predicted == review["label"],
            "formato_valido": predicted in ("positivo", "negativo", "neutro"),
            "tokens_input": response.usage.input_tokens,
            "tokens_output": response.usage.output_tokens,
        })
    return results


# ─── 4. Métricas ────────────────────────────────────────────────────────────

def compute_metrics(results: list[dict]) -> dict:
    """Calcula accuracy, consistencia de formato y tokens."""
    n = len(results)
    accuracy = sum(r["correcto"] for r in results) / n
    format_consistency = sum(r["formato_valido"] for r in results) / n
    avg_input_tokens = sum(r["tokens_input"] for r in results) / n
    avg_output_tokens = sum(r["tokens_output"] for r in results) / n
    return {
        "accuracy": accuracy,
        "format_consistency": format_consistency,
        "avg_input_tokens": avg_input_tokens,
        "avg_output_tokens": avg_output_tokens,
        "total_input_tokens": sum(r["tokens_input"] for r in results),
    }


# ─── 5. Impresión de tabla ──────────────────────────────────────────────────

def print_results_table(variant_name: str, results: list[dict], metrics: dict):
    """Imprime tabla detallada de resultados."""
    print(f"\n{'═' * 70}")
    print(f"  {variant_name}")
    print(f"{'═' * 70}")
    print(f"  {'Reseña (extracto)':<40} {'Real':<12} {'Predicción':<12} {'OK'}")
    print(f"  {'-' * 66}")
    for r in results:
        ok = "✓" if r["correcto"] else "✗"
        print(f"  {r['text']:<40} {r['label_real']:<12} {r['prediccion']:<12} {ok}")
    print(f"\n  Accuracy:            {metrics['accuracy']:.0%}")
    print(f"  Consistencia formato: {metrics['format_consistency']:.0%}")
    print(f"  Tokens input (prom):  {metrics['avg_input_tokens']:.0f}")
    print(f"  Tokens output (prom): {metrics['avg_output_tokens']:.1f}")
    print(f"  Tokens input total:   {metrics['total_input_tokens']}")


def print_comparison_table(all_metrics: dict):
    """Imprime tabla comparativa final."""
    print(f"\n{'═' * 70}")
    print("  TABLA COMPARATIVA")
    print(f"{'═' * 70}")
    print(f"  {'Variante':<20} {'Accuracy':>10} {'Formato':>10} {'Tokens/call':>12} {'Total tokens':>14}")
    print(f"  {'-' * 66}")
    for name, m in all_metrics.items():
        print(
            f"  {name:<20} {m['accuracy']:>9.0%} {m['format_consistency']:>9.0%} "
            f"{m['avg_input_tokens']:>11.0f} {m['total_input_tokens']:>13}"
        )
    print(f"\n  Nota: 'Tokens/call' = promedio de tokens de input por clasificación")
    print(f"  La diferencia refleja el overhead de los ejemplos few-shot.")


# ─── 6. Main ────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    client = anthropic.Anthropic()

    variants = [
        ("0-shot (sin ejemplos)", []),
        ("3-shot (3 ejemplos)",   EXAMPLES_3_SHOT),
        ("6-shot (6 ejemplos)",   EXAMPLES_6_SHOT),
    ]

    all_metrics = {}

    for name, examples in variants:
        system_prompt = build_system_prompt(examples)
        results = classify_reviews(client, system_prompt, REVIEWS)
        metrics = compute_metrics(results)
        print_results_table(name, results, metrics)
        all_metrics[name] = metrics

    print_comparison_table(all_metrics)

    # Nota sobre costo de los ejemplos
    base_tokens = all_metrics["0-shot (sin ejemplos)"]["avg_input_tokens"]
    tokens_3 = all_metrics["3-shot (3 ejemplos)"]["avg_input_tokens"]
    tokens_6 = all_metrics["6-shot (6 ejemplos)"]["avg_input_tokens"]
    print(f"\n  Overhead de tokens por ejemplos:")
    print(f"    3-shot vs 0-shot: +{tokens_3 - base_tokens:.0f} tokens/llamada")
    print(f"    6-shot vs 0-shot: +{tokens_6 - base_tokens:.0f} tokens/llamada")
    print(f"\n  Observación: si accuracy no mejora proporcionalmente al overhead,")
    print(f"  los ejemplos adicionales no justifican su costo en este modelo/tarea.")

File: typescript/04-prompts/few-shot.ts
// Comparación 0-shot vs 3-shot vs 6-shot en clasificación de sentimiento.
//
// Demuestra:
// - Cómo el número de ejemplos afecta accuracy y consistencia del formato
// - Majority label bias: con ejemplos desbalanceados, el modelo predice la clase dominante
// - Tokens consumidos por cada variante

// Cómo ejecutar: make ts SCRIPT=typescript/04-prompts/few-shot.ts

import Anthropic from "@anthropic-ai/sdk";

const MODEL = process.env["MODEL"] ?? "claude-sonnet-4-6";

// ─── 1. Dataset de prueba ────────────────────────────────────────────────────

interface Review {
  text: string;
  label: "positivo" | "negativo" | "neutro";
}

const REVIEWS: Review[] = [
  {
    text: "El producto llegó perfectamente empaquetado y funciona exactamente como se describe. Muy satisfecho.",
    label: "positivo",
  },
  {
    text: "Terrible experiencia. El artículo llegó roto y el servicio de atención al cliente no respondió.",
    label: "negativo",
  },
  {
    text: "El producto cumple lo básico. No es el mejor que he usado pero tampoco es malo. Precio razonable.",
    label: "neutro",
  },
  {
    text: "¡Increíble calidad! Superó todas mis expectativas. Lo recomendaría sin dudar.",
    label: "positivo",
  },
  {
    text: "Llegó tarde y el embalaje estaba dañado. El producto funciona pero la experiencia de compra fue mala.",
    label: "negativo",
  },
];

// ─── 2. Ejemplos few-shot ────────────────────────────────────────────────────

const EXAMPLES_3_SHOT: Review[] = [
  { text: "Excelente producto, muy buena calidad y entrega rápida.", label: "positivo" },
  { text: "No me gustó nada. Tuve que devolverlo al día siguiente.", label: "negativo" },
  { text: "Hace lo que promete. Ni más ni menos.", label: "neutro" },
];

const EXAMPLES_6_SHOT: Review[] = [
  ...EXAMPLES_3_SHOT,
  { text: "Mejor compra del año, totalmente recomendado para todos.", label: "positivo" },
  { text: "Producto defectuoso. Una pérdida de dinero total.", label: "negativo" },
  { text: "Está bien para lo que cuesta. No hay mucho que decir.", label: "neutro" },
];

// ─── 3. Construcción de prompts ──────────────────────────────────────────────

function buildExamplesBlock(examples: Review[]): string {
  const lines = ["<examples>"];
  for (const ex of examples) {
    lines.push("  <example>");
    lines.push(`    <texto>${ex.text}</texto>`);
    lines.push(`    <sentimiento>${ex.label}</sentimiento>`);
    lines.push("  </example>");
  }
  lines.push("</examples>");
  return lines.join("\n");
}

function buildSystemPrompt(examples: Review[]): string {
  const base =
    "Clasifica el sentimiento de reseñas de producto como: positivo, negativo o neutro.\n" +
    "Responde SOLO con una de estas tres palabras: positivo, negativo, neutro.\n" +
    "Sin explicaciones adicionales.\n";
  if (examples.length === 0) return base;
  return base + "\n" + buildExamplesBlock(examples);
}

// ─── 4. Clasificación ────────────────────────────────────────────────────────

interface ClassificationResult {
  text: string;
  labelReal: string;
  prediccion: string;
  correcto: boolean;
  formatoValido: boolean;
  tokensInput: number;
  tokensOutput: number;
}

async function classifyReviews(
  client: Anthropic,
  systemPrompt: string,
  reviews: Review[]
): Promise<ClassificationResult[]> {
  const results: ClassificationResult[] = [];
  for (const review of reviews) {
    const response = await client.messages.create({
      model: MODEL,
      max_tokens: 20,
      system: systemPrompt,
      messages: [{ role: "user", content: review.text }],
    });

    const rawOutput = (response.content[0] as Anthropic.TextBlock).text.trim().toLowerCase();
    const match = rawOutput.match(/\b(positivo|negativo|neutro)\b/);
    const predicted = match ? match[1] : rawOutput;
    const validLabels = ["positivo", "negativo", "neutro"];

    results.push({
      text: review.text.slice(0, 50) + "...",
      labelReal: review.label,
      prediccion: predicted,
      correcto: predicted === review.label,
      formatoValido: validLabels.includes(predicted),
      tokensInput: response.usage.input_tokens,
      tokensOutput: response.usage.output_tokens,
    });
  }
  return results;
}

// ─── 5. Métricas ─────────────────────────────────────────────────────────────

interface Metrics {
  accuracy: number;
  formatConsistency: number;
  avgInputTokens: number;
  avgOutputTokens: number;
  totalInputTokens: number;
}

function computeMetrics(results: ClassificationResult[]): Metrics {
  const n = results.length;
  return {
    accuracy: results.filter((r) => r.correcto).length / n,
    formatConsistency: results.filter((r) => r.formatoValido).length / n,
    avgInputTokens: results.reduce((s, r) => s + r.tokensInput, 0) / n,
    avgOutputTokens: results.reduce((s, r) => s + r.tokensOutput, 0) / n,
    totalInputTokens: results.reduce((s, r) => s + r.tokensInput, 0),
  };
}

// ─── 6. Impresión de resultados ──────────────────────────────────────────────

function printResultsTable(variantName: string, results: ClassificationResult[], metrics: Metrics): void {
  console.log(`\n${"═".repeat(70)}`);
  console.log(`  ${variantName}`);
  console.log(`${"═".repeat(70)}`);
  console.log(`  ${"Reseña (extracto)".padEnd(40)} ${"Real".padEnd(12)} ${"Predicción".padEnd(12)} OK`);
  console.log(`  ${"-".repeat(66)}`);
  for (const r of results) {
    const ok = r.correcto ? "✓" : "✗";
    console.log(`  ${r.text.padEnd(40)} ${r.labelReal.padEnd(12)} ${r.prediccion.padEnd(12)} ${ok}`);
  }
  console.log(`\n  Accuracy:             ${(metrics.accuracy * 100).toFixed(0)}%`);
  console.log(`  Consistencia formato: ${(metrics.formatConsistency * 100).toFixed(0)}%`);
  console.log(`  Tokens input (prom):  ${metrics.avgInputTokens.toFixed(0)}`);
  console.log(`  Tokens output (prom): ${metrics.avgOutputTokens.toFixed(1)}`);
  console.log(`  Tokens input total:   ${metrics.totalInputTokens}`);
}

function printComparisonTable(allMetrics: Record<string, Metrics>): void {
  console.log(`\n${"═".repeat(70)}`);
  console.log("  TABLA COMPARATIVA");
  console.log(`${"═".repeat(70)}`);
  console.log(
    `  ${"Variante".padEnd(20)} ${"Accuracy".padStart(10)} ${"Formato".padStart(10)} ${"Tokens/call".padStart(12)} ${"Total tokens".padStart(14)}`
  );
  console.log(`  ${"-".repeat(68)}`);
  for (const [name, m] of Object.entries(allMetrics)) {
    console.log(
      `  ${name.padEnd(20)} ${(m.accuracy * 100).toFixed(0).padStart(9)}% ` +
      `${(m.formatConsistency * 100).toFixed(0).padStart(9)}% ` +
      `${m.avgInputTokens.toFixed(0).padStart(11)} ` +
      `${m.totalInputTokens.toString().padStart(13)}`
    );
  }
  console.log(`\n  Nota: 'Tokens/call' = promedio de tokens de input por clasificación`);
}

// ─── 7. Main ──────────────────────────────────────────────────────────────────

async function main(): Promise<void> {
  const client = new Anthropic();

  const variants: [string, Review[]][] = [
    ["0-shot (sin ejemplos)", []],
    ["3-shot (3 ejemplos)", EXAMPLES_3_SHOT],
    ["6-shot (6 ejemplos)", EXAMPLES_6_SHOT],
  ];

  const allMetrics: Record<string, Metrics> = {};

  for (const [name, examples] of variants) {
    const systemPrompt = buildSystemPrompt(examples);
    const results = await classifyReviews(client, systemPrompt, REVIEWS);
    const metrics = computeMetrics(results);
    printResultsTable(name, results, metrics);
    allMetrics[name] = metrics;
  }

  printComparisonTable(allMetrics);

  const base = allMetrics["0-shot (sin ejemplos)"].avgInputTokens;
  const t3 = allMetrics["3-shot (3 ejemplos)"].avgInputTokens;
  const t6 = allMetrics["6-shot (6 ejemplos)"].avgInputTokens;
  console.log(`\n  Overhead de tokens por ejemplos:`);
  console.log(`    3-shot vs 0-shot: +${(t3 - base).toFixed(0)} tokens/llamada`);
  console.log(`    6-shot vs 0-shot: +${(t6 - base).toFixed(0)} tokens/llamada`);
}

main().catch(console.error);
